gbn/timeout_manager.go

```diff
@@ -0,0 +1,327 @@
+	// sending a FIN message will be time out.
+	finSendTimeout time.Duration
+
+	// sendTimeout defines the max time we will wait to send a msg before
+	// timing out.
+	sendTimeout time.Duration
+
+	// recvTimeout defines the max time we will wait to receive a msg before
+	// timing out.
+	recvTimeout time.Duration
+
+	// pingTime represents at which frequency we will send pings to the
+	// counterparty if we've received no packet.
+	pingTime time.Duration
+
+	// pongTime represents how long we will wait for the expect a pong
+	// response after we've sent a ping. If no response is received within
+	// the time limit, we will close the connection.
+	pongTime time.Duration
+
+	// timeoutManagerMu should be locked when updating or accessing any of
+	// timeout manager's timeout fields.
+	timeoutManagerMu sync.RWMutex
+
+	// responseCounter represents the current number of corresponding
+	// responses received since last updating the resend timeout.
+	responseCounter int
+
+	// timeoutUpdateFrequency represents the frequency of how many
+	// corresponding responses we need to receive until the resend timeout
+	// will be updated.
+	timeoutUpdateFrequency int
+
+	sentTimes   map[uint8]time.Time
+	sentTimesMu sync.Mutex
+}
+
+// NewTimeOutManager creates a new timeout manager.
+func NewTimeOutManager(isServer bool) *TimeoutManager {
+	return &TimeoutManager{
+		isServer:               isServer,
+		resendTimeout:          defaultResendTimeout,
+		handshakeTimeout:       defaultHandshakeTimeout,
+		useStaticTimeout:       false,
+		resendMultiplier:       defaultResendMultiplier,
+		finSendTimeout:         defaultFinSendTimeout,
+		recvTimeout:            DefaultRecvTimeout,
+		sendTimeout:            DefaultSendTimeout,
+		sentTimes:              make(map[uint8]time.Time),
+		timeoutUpdateFrequency: defaultTimeoutUpdateFrequency,
+	}
+}
+
+// Sent should be called when a message is sent by the connection. The resent
+// parameter should be set to true if the message is a resent message.
+func (m *TimeoutManager) Sent(msg Message, resent bool) {
+	if m.useStaticTimeout {
+		return
+	}
+
+	// We will dynamically update the resend timeout throughout the lifetime
+	// of the connection, to ensure that it reflects the current response
+	// time. Therefore we'll keep track of when we sent a package, and when
+	// we receive the corresponding response.
+	// If we're resending a message, we can't know if a corresponding
+	// response is the response to the resent message, or the original
+	// message. Therefore, we never update the resend timeout after
+	// resending a message.
+	switch msg := msg.(type) {
+	case *PacketSYN:
+		if resent {
+			// If we've resent the SYN, we'll reset the
+			// latestSentSYNTime to the zero value, to ensure
+			// that we don't update the resend timeout based on
+			// the corresponding response, as we can't know if
+			// the response is for the resent SYN or the original
+			// SYN.
+			m.latestSentSYNTime = time.Time{}
```

Comment:
Do we need to guard this by a mutex as well? Since it looks to me that `Sent()` and `Received()` could be called concurrently?

Reply:
Same for the `m.useStaticTimeout`. I guess we guard the _write_ access to that but not the read access, so it could still cause a concurrent access issue.

Reply:
> Do we need to guard this by a mutex as well? Since it looks to me that `Sent()` and `Received()` could be called concurrently?

Ah yeah, great catch! I've addressed this in the latest push.

> Same for the `m.useStaticTimeout`. I guess we guard the _write_ access to that but not the read access, so it could still cause a concurrent access issue.

Since the `useStaticTimeout` is only set through the functional options that's passed with the `TimeoutManager`'s constructor, this shouldn't be an issue.

Reply:
I've now removed the `SetStaticResendTimout` function!